### PR TITLE
Support scoped node modules

### PR DESCRIPTION
**Pull request by jamesblight**
Rename scoped modules prefixed with `@` so they are not picked up
by Bazel as illegal labels. Don't treat scoped module parent directories
as node modules.

### CODE REVIEW

File: node/internal/yarn_repository.bzl [hunk 1]
```diff
@@ -28,8 +28,17 @@ def _yarn_repository_impl(ctx):
...
     modules = []
+    # Rename and move scoped modules with invalid bazel labels
+    for module in modules_path.readdir():
+      if module.basename.startswith("@"):
+        for scoped_module in module.readdir():
+          execute(ctx, ["mv", scoped_module, "%s/%s" % (modules_path, mangle_package_name(module.basename, scoped_module.basename))])
+
     for module in modules_path.readdir():
         if module.basename.startswith("."): continue
+        # scoped modules just contain other modules so don't include them
```
Comment (mterring) [inline node/internal/yarn_repository.bzl:39]: This check appears to be unnecessary as the rename will take care of it?

File: node/internal/npm_repository.bzl [hunk 2]
```diff
@@ -35,8 +35,19 @@ def _npm_repository_impl(ctx):
...
+    # Rename and move scoped modules with invalid bazel labels
+    for module in modules_path.readdir():
+      if module.basename.startswith("@"):
+        for scoped_module in module.readdir():
+          execute(ctx, ["mv", scoped_module, "%s/%s" % (modules_path, mangle_package_name(module.basename, scoped_module.basename))])
+
+
     for module in modules_path.readdir():
         if module.basename.startswith("."): continue
+        # scoped modules just contain other modules so don't include them
```
Comment (mterring) [inline node/internal/npm_repository.bzl:48]: This check appears to be unnecessary as the rename will take care of it?

File: node/internal/node_utils.bzl [hunk 3]
```diff
@@ -118,31 +135,41 @@ def node_install(ctx, install_path, modules):
...
                     ]
-                    if w2.wrapped_deps: fail("nested wrapped dependencies not supported deeper than 3 levels (in %s)" % w3.label)
+                    for w4 in w3.wrapped_deps:
+                        inputs += [w4.file]
+                        w4Name = demangle_package_name(w4.name)
+                        cmds += [
+                            "mkdir -p %s/%s/node_modules/%s/node_modules/%s/node_modules/%s/node_modules/%s" % (install_path.path, mName, wName, w2Name, w3Name, w4Name),
+                            "tar -xzf %s -C %s/%s/node_modules/%s/node_modules/%s/node_modules/%s/node_modules/%s --strip-components 1" % (w4.file.path, install_path.path, mName, wName, w2Name, w3Name, w4Name),
+                        ]
+                    if w3.wrapped_deps: fail("nested wrapped dependencies not supported deeper than 4 levels (in %s)" % w4.label)
```
Comment (mterring) [inline node/internal/node_utils.bzl]: Suspect this should be w4.wrapped_deps - existing bug had w2... Maybe going to 4 levels wasn't required?

File: node/internal/node_utils.bzl [hunk 4]
```diff
@@ -17,27 +17,44 @@ node_library(
...
         deps = execute(repository_ctx, deps_cmd).stdout.strip()
+
         indeps = ["//%s:node_module" % (i) for i in repository_ctx.attr.indeps.get(t.module.basename, [])]
 
         wrapped_deps = []
         wrapped_path = t.module.get_child("node_modules")
         if wrapped_path.exists:
             for sub_module in wrapped_path.readdir():
                 if sub_module.basename.startswith("."): continue
+                # scoped modules aren't actual modules so don't include them
```
Comment (mterring) [inline node/internal/node_utils.bzl]: Does this mean we don't support scoped modules as wrapped dependencies?

File: node/internal/node_utils.bzl [hunk 5]
```diff
@@ -17,27 +17,44 @@ node_library(
...
     ]).stdout.split("\n")
     repo_prefix_len = len(str(repository_ctx.path(".")))
+
     for w in [repository_ctx.path(w[repo_prefix_len+1:]) for w in wrapped_paths if w and not w.endswith(".bin")]:
-        targets.append(struct(module = w, path = str(w)[repo_prefix_len+14:]))
+        if w.basename.startswith("@"):
+            for scoped_module in w.readdir():
+                mangled_package_name = mangle_package_name(w.basename, scoped_module.basename)
+                mangled_module_path = "%s/%s" % (w.dirname, mangled_package_name)
+                # mangled_module_path = "%s/%s__SLASH__%s" % (w.dirname, w.basename.replace("@", "__AT__"), scoped_module.basename)
```
Comment (mterring) [inline node/internal/node_utils.bzl]: Remove commented code

File: node/internal/node_utils.bzl [hunk 6]
```diff
@@ -17,27 +17,44 @@ node_library(
...
+                execute(repository_ctx, ["mv", scoped_module, mangled_module_path])
+                renamed_module = repository_ctx.path(w.dirname).get_child(mangled_package_name)
+                # renamed_module = repository_ctx.path(w.dirname).get_child("%s__SLASH__%s" % (w.basename.replace("@", "__AT__"), scoped_module.basename))
```
Comment (mterring) [inline node/internal/node_utils.bzl]: Remove commented code

File: node/internal/node_library.bzl [hunk 7]
```diff
@@ -53,19 +53,25 @@ module_group = rule(
...
     for p in srcs_by_pkg:
+        if strip_prefix == "":
+            strip_prefix = "." if p == native.package_name() else "/%s" % p
+        else:
+            strip_prefix = "./%s" % strip_prefix if p == native.package_name() else "/%s/%s" % (p, strip_prefix)
+
         pkg_tar(
             name = "%s-%s" % (name, p.replace("/", "_")),
-            strip_prefix = "." if p == native.package_name() else "/%s" % p,
+            strip_prefix = strip_prefix if p == native.package_name() else "/%s" % p,
```
Comment (mterring) [inline node/internal/node_library.bzl]: No need for the if...

**Review by mterring — CHANGES_REQUESTED** [review]

**Review by mterring — CHANGES_REQUESTED** [review]
Also need to extend the examples to cover scoped modules

File: node/internal/node_utils.bzl (hunk 3 above)
Comment (jamesblight) [reply node/internal/node_utils.bzl]: I think you're right, I just needed to change it to `w3.wrapped_deps` in the original code. Should we keep support for 4 levels?

File: node/internal/yarn_repository.bzl (hunk 1 above)
Comment (jamesblight) [reply node/internal/yarn_repository.bzl:39]: The scoped modules are moved out of the parent directory but the parent - with the `@` prefix - remains and needs to be skipped. (It's empty)

File: node/internal/node_utils.bzl (hunk 4 above)
Comment (jamesblight) [reply node/internal/node_utils.bzl]: No. We're just skipping over the directory that contained them before they were renamed and removed. [Here](https://github.com/happy-co/rules_node/pull/1/files/86a6cb6a129b0288bed0ceeec51ac344725b9dea#diff-967bff59cbc0d8ceaeacdc8d8d03f7f6R28)

**Review by mterring — APPROVED** [review]

**Pull request merged**